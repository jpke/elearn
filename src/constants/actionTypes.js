export const LOADING = 'LOADING';
export const BAD_RESPONSE = 'BAD_RESPONSE';
export const LOG_IN = 'LOG_IN';
export const LOG_OUT = 'LOG_OUT';

export const SELECT_COURSE = 'SELECT_COURSE';
export const EDIT_COURSE = 'EDIT_COURSE';
export const ADD_USER = 'ADD_USER';
export const UPDATE_ENROLLABLE = 'UPDATE_ENROLLABLE';
export const UPDATE_ENROLLED = 'UPDATE_ENROLLED';
export const UPDATE_COURSE = 'UPDATE_COURSE';

export const LOAD_QUIZ = 'LOAD_QUIZ';
export const TOGGLE_QUIZ_VIEW = 'TOGGLE_QUIZ_VIEW';

export const CREATE_QUIZ = 'CREATE_QUIZ';
export const UPDATE_QUIZ = 'UPDATE_QUIZ';
export const DELETE_QUIZ = 'DELETE_QUIZ';
export const SAVE_QUIZ = 'SAVE_QUIZ';

export const START_QUIZ = 'START_QUIZ';
export const SELECT_ANSWER = 'SELECT_ANSWER';
export const NEXT_QUESTION = 'NEXT_QUESTION';
export const PREVIOUS_QUESTION = 'PREVIOUS_QUESTION';
export const SUBMIT_QUIZ = 'SUBMIT_QUIZ';
export const GET_LESSONS = 'GET_LESSONS';
export const GET_PDF = 'GET_PDF';
