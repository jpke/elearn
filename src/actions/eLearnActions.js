// import cookie from 'react-cookie'
import * as types from '../constants/actionTypes';

export function loading(item) {
  return {
    type: 'LOADING',
    item
  };
}

function loggedIn(response) {
  console.log("response: ", response);
  localStorage.setItem("login", {
    userName: response.name,
    user_Id: response._id,
    courses: response.courses,
    token: response.token
  });
  return {
    type: types.LOG_IN,
    userName: response.name,
    user_Id: response._id,
    courses: response.courses,
    token: response.token
  };
}

export function register(userName, email, password) {
  return function (dispatch) {
    dispatch(loading('register'));
    try {
      fetch('http://localhost:8080/elearn/users', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json'
        },
        body: JSON.stringify({
          name: userName,
          email: email,
          password: password
        })
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          if(response.body.message === 'email already associated with an account') {
            return dispatch({
              type: types.EMAIL_USED,
              email
            });
          } else {
            let error = response;
            throw error;
          }
        }
        return response.json()
      })
      .then(response => {
        window.localStorage.loggedIn = true;
        dispatch(loading(''));
        dispatch(loggedIn(response));
      })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}

export function logIn(email, password) {
  return function (dispatch) {
    dispatch(loading('logIn'));
    try {
      fetch('http://localhost:8080/elearn/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json'
        },
        withCredentials: true,
        body: JSON.stringify({
          email: email,
          password: password
        })
      })
      .then(response => {
        if(response.status != 302) {
          let error = response;
          throw error;
        }
        return response.json()
      })
      .then(response => {
        window.localStorage.loggedIn = true;
        // console.log("set-cookie: ", response)
        // cookie.save("tokens",
        //   response.token,
        //   {
        //     httpOnly: true
        //   });
        dispatch(loading(''));
        dispatch(loggedIn(response));
      })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}

export function selectCourse(courseName, course_Id) {
  return {
    type: types.SELECT_COURSE,
    courseName,
    course_Id
  };
}

export function logOut() {
  //consider blacklisting token serverside in future
  window.localStorage.loggedIn = false;
  // cookie.remove('token');
  return {
    type: types.LOG_OUT
  };
}

export function startQuiz(token, course_Id) {
  return function (dispatch) {
    dispatch(loading('startQuiz'));
    //pull down quiz questions, then
    try {
      fetch('http://localhost:8080/elearn/quiz/'.concat(course_Id), {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        }
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          let error = response;
          throw error;
        }
        return response.json()
      })
      .then(response => {
        dispatch(loading(''));
        console.log("quiz actions: ", response);
        return dispatch({
          type: types.START_QUIZ,
          quizData: response
        });
        })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}

export function selectAnswer(answerSelected, idSelected, item) {
  return {
    type: types.SELECT_ANSWER,
    answerSelected,
    idSelected,
    itemSelected: parseInt(item)
  };
}

export function nextQuestion() {
  return {
    type: types.NEXT_QUESTION
  };
}

export function prevQuestion() {
  return {
    type: types.PREVIOUS_QUESTION
  };
}

export function submitQuiz(quizData, quizTitle, quizId, _id, token) {
  return function (dispatch) {
    dispatch(loading('submitQuiz'));
    try {
      fetch('http://localhost:8080/elearn/quiz/submit', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        body: JSON.stringify({
          title: quizTitle,
          quizData: quizData,
          quiz_Id: quizId,
          user_Id: _id
        })
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          let error = response;
          throw error;
        } else {
          return response;
        }
      })
      .then(response => response.json())
      .then((response) => {
        console.log("action response: ", response);
        dispatch(loading(''));
        return dispatch({
          type: types.SUBMIT_QUIZ,
          score: response.score
        });
        })
    } catch(error) {
      console.log("error response: ", error);
    }
  }
}

export function getLessons(token) {
  return function (dispatch) {
    // let cookieToken = cookie.load('token');
    // console.log("cookieToken: ", cookieToken);
    dispatch(loading('lessons'));
    try {
      fetch('http://localhost:8080/elearn/lessons', {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        // withCredentials: true,
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          let error = response;
          throw error;
        }
        return response.json()
      })
      .then(response => {
        dispatch(loading(''));
        return dispatch({
          type: types.GET_LESSONS,
          lessons: response.entries
        });
        })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}

export function getPDF(pdfId, token) {
  return function (dispatch) {
    dispatch(loading("pdf"));
    try {
      fetch('http://localhost:8080/elearn/lessons/'.concat(pdfId), {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        // withCredentials: true,
      })
      .then(response => {
        if(response.status < 200 || response.status >= 300) {
          let error = response;
          throw error;
        }
        return response.json()
      })
      .then(response => {
        dispatch(loading(""));
        console.log("response: ", response);
          return dispatch({
            type: types.GET_PDF,
            response
          });
        })
    } catch(error) {
      console.log("error response: ", error);
    }
  };
}
