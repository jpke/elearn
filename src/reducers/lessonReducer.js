import * as types from '../constants/actionTypes';

const initialState = {
  lessons: [],
  selectedPdf: {}
};

export default function lessonReducer(state = initialState, action) {
  switch(action.type) {
    case types.GET_LESSONS:
      return {
        ...state,
        lessons: action.lessons
      };
    case types.GET_PDF:
    console.log('reducer response: ',action.response);
      return {
        ...state,
        selectedPdf: action.response.selectedPdf
      }
    default:
    return state;
  }
}
