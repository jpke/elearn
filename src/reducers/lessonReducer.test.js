/* eslint-disable no-undef */
/* eslint-disable import/no-named-as-default */

import React from 'react';
import lessonReducer from './lessonReducer';
import {initialState} from './lessonReducer';
import * as types from '../constants/actionTypes';
import * as actions from '../actions/eLearnActions';

let state;
beforeEach(() => {
  state = JSON.parse(JSON.stringify(initialState));
})

it('should handle GET_LESSONS', () => {
  expect(lessonReducer(state, {
    type: types.GET_LESSONS,
    lessons: [
      {
        type: "file",
        id: "0000",
        etag: "1",
        name: "Example Lesson File.pdf",
        modified_at: "2017-02-02T12:29:25-08:00",
        size: 15342
      },
      {
        type: "file",
        id: "0001",
        etag: "1",
        name: "Example Lesson File Two.pdf",
        modified_at: "2017-02-02T12:29:25-08:00",
        size: 15342
      },
      {
        type: "file",
        id: "0002",
        etag: "1",
        name: "Example Lesson File Three.pdf",
        modified_at: "2017-02-02T12:29:25-08:00",
        size: 15342
      }
    ]
  }
  )).toMatchSnapshot();
})

it('should handle GET_PDF', () => {
  state.lessons = [
    {
      type: "file",
      id: "0000",
      etag: "1",
      name: "Example Lesson File.pdf",
      modified_at: "2017-02-02T12:29:25-08:00",
      size: 15342
    },
    {
      type: "file",
      id: "0001",
      etag: "1",
      name: "Example Lesson File Two.pdf",
      modified_at: "2017-02-02T12:29:25-08:00",
      size: 15342
    },
    {
      type: "file",
      id: "0002",
      etag: "1",
      name: "Example Lesson File Three.pdf",
      modified_at: "2017-02-02T12:29:25-08:00",
      size: 15342
    }
  ];
  expect(lessonReducer(state, {
    type: types.GET_LESSONS,
    selectedPdf: {
      preview: "https://somePreviewAddress.com",
      download: "https://someDownloadAddress.com",
      name: "Example Lesson File.pdf",
      id: "0000"
    }
  }
  )).toMatchSnapshot();
})
