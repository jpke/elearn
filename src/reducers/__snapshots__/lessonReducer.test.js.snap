exports[`test should handle GET_LESSONS 1`] = `
Object {
  "lessons": Array [
    Object {
      "etag": "1",
      "id": "0000",
      "modified_at": "2017-02-02T12:29:25-08:00",
      "name": "Example Lesson File.pdf",
      "size": 15342,
      "type": "file",
    },
    Object {
      "etag": "1",
      "id": "0001",
      "modified_at": "2017-02-02T12:29:25-08:00",
      "name": "Example Lesson File Two.pdf",
      "size": 15342,
      "type": "file",
    },
    Object {
      "etag": "1",
      "id": "0002",
      "modified_at": "2017-02-02T12:29:25-08:00",
      "name": "Example Lesson File Three.pdf",
      "size": 15342,
      "type": "file",
    },
  ],
  "selectedPdf": Object {},
}
`;

exports[`test should handle GET_PDF 1`] = `
Object {
  "lessons": undefined,
  "selectedPdf": Object {},
}
`;
